### PR TITLE
[SCEV] Support ule/sle exit counts via widening

### PR DESCRIPTION
**Pull request by nikic**
If we have an exit condition of the form IV <= Limit, we will first try to convert it into IV < Limit+1 or IV-1 < Limit based on range info (in icmp simplification). If that fails, we try to convert it to IV < Limit + 1 based on controlling exits in non-infinite loops.

However, if all else fails, we can still determine the exit count by rewriting to ext(IV) < ext(Limit) + 1, where the zero/sign extension ensures that the addition does not overflow.

Proof: https://alive2.llvm.org/ce/z/iR-iYd

### CODE REVIEW

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 1]
```diff
@@ -9198,8 +9198,21 @@ ScalarEvolution::ExitLimit ScalarEvolution::computeExitLimitFromICmp(
...
     if (!EnableFiniteLoopControl || !ControllingFiniteLoop ||
-        !isLoopInvariant(RHS, L))
-      break;
+        !isLoopInvariant(RHS, L)) {
+      // Otherwise, perform the addition in a wider type, to avoid overflow.
+      auto *OldType = dyn_cast<IntegerType>(LHS->getType());
+      if (!OldType)
+        break;
+      auto *NewType =
+          Type::getIntNTy(OldType->getContext(), OldType->getBitWidth() * 2);
```
Comment (nikic) [inline llvm/lib/Analysis/ScalarEvolution.cpp:9207]: Just adding +1 would be sufficient here, but I've opted to go for *2 which will likely result in friendlier IR if it gets expanded.

**Comment by efriedma-quic** [comment]
In general, you end up with a sext/zext of some expression which still isn't analyzable... but I guess you end up with a useful result in a lot of cases, as demonstrated by the testcases.  Maybe worth explicitly noting that you expect the sign/zero-extension to simplify.

My biggest concern here is that we'll somehow end up generating code for a bunch of double-register-width math, which would be relatively expensive to compute.  Not sure how likely that is, though.

**Comment by nikic** [comment]
> In general, you end up with a sext/zext of some expression which still isn't analyzable... but I guess you end up with a useful result in a lot of cases, as demonstrated by the testcases. Maybe worth explicitly noting that you expect the sign/zero-extension to simplify.

Yes, this will mainly do something useful is the LHS is an addrec with the corresponding nuw or nsw flag, in which case the extension is sunk into the addrec and we get an analyzable condition.

Though based on my testing, I think this change has very little practical impact. Probably non-strict exit conditions that can't be converted to strict ones are pretty rare. I mainly need this to avoid a regression in one PhaseOrdering test for an upcoming change.
 
> My biggest concern here is that we'll somehow end up generating code for a bunch of double-register-width math, which would be relatively expensive to compute. Not sure how likely that is, though.

isHighCostExpansion does cost modelling based on TTI, so I'd expect we would mostly refuse to expand cases where the doubling goes beyond the register size.

**Comment by nikic** [comment]
ping

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

IIUC the more expensive expansions should be taken care of mostly by `isHighCostExpansion` for cases where we now compute more complicated BTCs (vs not being able to compute them before).

Would still be good to wait a day or so in case @efriedma-quic has additional thoughts

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Pull request merged**